INDI/INDI/indiserver.c

```diff
@@ -1948,150 +1958,211 @@ static int readFromDriver(DvrInfo *dp)
+}
 
+/* Setup for messages to or from local clients */
+static int setup_client_messages(int maxfd, fd_set* prs, fd_set* pws)
+{
+    int i;
     /* add all client readers and client writers with work to send */
     for (i = 0; i < nclinfo; i++)
     {
-        ClInfo *cp = &clinfo[i];
-        if (cp->active)
-        {
-            FD_SET(cp->s, &rs);
-            if (nFQ(cp->msgq) > 0)
-                FD_SET(cp->s, &ws);
-            if (cp->s > maxfd)
-                maxfd = cp->s;
-        }
+        ClInfo* cp = clinfo + i;
+        if (!cp->active) { continue; }
+        FD_SET(cp->s, prs);
+        if (nFQ(cp->msgq) > 0) { FD_SET(cp->s, pws); }
+        if (cp->s > maxfd) { maxfd = cp->s; }
     }
+    return maxfd;
+}
 
+static int setup_driver_messages(int maxfd, fd_set* prs, fd_set* pws)
+{
+    int i;
     /* add all driver readers and driver writers with work to send */
     for (i = 0; i < ndvrinfo; i++)
     {
-        DvrInfo *dp = &dvrinfo[i];
-        if (dp->active)
-        {
-            FD_SET(dp->rfd, &rs);
-            if (dp->rfd > maxfd)
-                maxfd = dp->rfd;
-#           if 0
-            if (dp->pid != REMOTEDVR)
-            {
-                FD_SET(dp->efd, &rs);
-                if (dp->efd > maxfd)
-                    maxfd = dp->efd;
-            }
-#           endif/*0*/
-            if (nFQ(dp->msgq) > 0)
-            {
-                FD_SET(dp->wfd, &ws);
-                if (dp->wfd > maxfd)
-                    maxfd = dp->wfd;
-            }
-        }
+        DvrInfo *dp = dvrinfo + i;
+        if (!dp->active || dp->restartDelayus > 0) { continue; }
+
+        FD_SET(dp->rfd, prs);
+        if (dp->rfd > maxfd) { maxfd = dp->rfd; }
+        if (nFQ(dp->msgq) < 1) { continue; }
+        FD_SET(dp->wfd, pws);
+        if (dp->wfd > maxfd) { maxfd = dp->wfd; }
     }
+    return maxfd;
+}
 
-    /* wait for action */
-    s = select(maxfd + 1, &rs, &ws, NULL, NULL);
-    if (s < 0)
+void handle_restart_list(struct timeval* ptv)
+{
+    /* Calculate time spent in select(2) call from remaining time in tv;
+     * ensure time spend is at least 1us
+     */
+    int time_in_select = ((SELECT_WAITs - ptv->tv_sec) * Mus) - ptv->tv_usec;
+    time_in_select = time_in_select > 0 ? time_in_select : 1;
+
+    /* Loop over linked list of drivers to restart */
+    pDvr pdvr;
+    for (pdvr = pRestarts; pdvr; pdvr = pdvr->pNextToRestart)
     {
-        if(errno == EINTR)
-            return;
-        fprintf(stderr, "%s: select(%d): %s\n", indi_tstamp(NULL), maxfd + 1, strerror(errno));
-        Bye();
+        /* Reduce remaining delay by time spent in select(2) call
+         * Do nothing more with this driver if delay has not expired
+         */
+        pdvr->restartDelayus -= time_in_select;
+        if (pdvr->restartDelayus > 0) continue;
+
+        /* Drop this driver from the to-be-restarted linked list */
+        pRestarts = pdvr->pNextToRestart;
+
+        startDvr(pdvr);
     }
+}
```

Comment:
This routine, handle_restart_list, calculates the duration of the select(2) call in indiRun below, decrements the DvrInfo restartDelayus value for all drivers that are waiting to restart. and attempts to restart the driver when its restartDelayus value reaches 0 or less.